### PR TITLE
Easier extending/replacing of key algorithms

### PR DESCRIPTION
**Pull request by friedcell**
Changed some code to make jwk algorithm implementations easily extendable.

If you want to replace a certain key implementation you only do `jwk.ALGORITHMS.register_key("[algorithm name]", [key class])` and from that moment on the algorithm will use a different class to do everything.

While doing it, made some stuff a bit more pythonic.

### CODE REVIEW

**Comment by mpdavis** [comment]
I'm curious what the use case for this is?  Are you looking to define and use custom algorithms at run time, or is there another reason?

**Comment by friedcell** [comment]
Correct, using a custom ES256 implementation on a RaspberryPI. Also kind-of resolves https://github.com/mpdavis/python-jose/issues/41 as it gets much easier to switch implementations.

**Comment by friedcell** [comment]
Code to switch an implementation (eg. customjose.py):
```python
from jose import jwk

class MyKeyImpl(jwk.Key):
    # custom implementation here

def register():
    jwk.ALGORITHMS.register_key("ES256", MyKeyImpl)
```

Where you use jose:
Before:
```python
from jose import jwt

jwk.encode({}, "private key", algorithm="ES256")
```
After:
```python
from customjose import register
from jose import jwt

register()

jwk.encode({}, "private key", algorithm="ES256")
```

The runtime switch is dead simple and you don't have to change any code in your application.

But the change also means that you can have multiple implementations of the Key classes and use `jose.constants` (I'd actually rename this to algorithms) to decide between them in `get_key()`. That way you can have multiple implementations in the package and switch between them depending on which library is installed.

**Comment by friedcell** [comment]
@mpdavis any feedback?

**Comment by zejn** [comment]
Hi,

To check the utility of this patch, I wrote a cryptography backend for ES256/ES384/ES512 in a separate repo at https://github.com/zejn/python-jose/commit/f4a840a2516954114edaac324a4933bc31b29c16.

There's still an unsettled way how to handle different backend crypto implementations. Currently used ecdsa is pure python and slow. There's a pull request for pyelliptic at https://github.com/mpdavis/python-jose/pull/39, but it doesn't work with other than SHA256 - https://github.com/yann2192/pyelliptic/pull/53 - and pyelliptic also does not support OpenSSL 1.1 - https://github.com/yann2192/pyelliptic/issues/50.

There's also an issue about supporting cryptography here - https://github.com/mpdavis/python-jose/issues/41.

I find this PR working, it touches very limited amount of code. Further rework would be welcome as it would allow to specify which key class to use if one does not want to use globally registered one.

File: jose/constants.py [hunk 1]
```diff
@@ -31,3 +32,28 @@ class ALGORITHMS(object):
         ES384: hashlib.sha384,
         ES512: hashlib.sha512,
     }
+
+    KEYS = {}
+
+    def get_key(self, algorithm):
+        from jose.jwk import HMACKey, RSAKey, ECKey
```
Comment (mpdavis) [inline jose/constants.py]: I'm not a fan of this circular dependency.  I would move this key registration/lookup logic to `jwk.py`.

File: jose/constants.py [hunk 2]
```diff
@@ -31,3 +32,28 @@ class ALGORITHMS(object):
...
+        return None
+
+    def register_key(self, algorithm, key_class):
+        from jose.jwk import Key
+        if issubclass(key_class, Key):
+            self.KEYS[algorithm] = key_class
+            self.SUPPORTED.add(algorithm)
+            return True
+        else:
+            return False
```
Comment (mpdavis) [inline jose/constants.py]: Attempting to register an invalid key should raise a `TypeError`

**Comment by mpdavis** [comment]
@friedcell My apologies for taking so long to get back to you on this.  I really appreciate the work.

I only have a couple of comments on the PR, and I think overall it will be a good addition.

**Comment by friedcell** [comment]
Updated the code per request

File: jose/constants.py [hunk 3]
```diff
@@ -31,3 +32,16 @@ class ALGORITHMS(object):
         ES384: hashlib.sha384,
         ES512: hashlib.sha512,
     }
+
+    KEYS = {}
+
+    def register_key(self, algorithm, key_class):
+        from jose.jwk import Key
```
Comment (mpdavis) [inline jose/constants.py]: This still has a wonky dependency.  Can you move it over as well?

**Comment by friedcell** [comment]
Moved that too

**Comment by friedcell** [comment]
@mpdavis ping?

**Comment by mpdavis** [comment]
LGTM

**Comment by mpdavis** [comment]
Released in [1.4.0](https://github.com/mpdavis/python-jose/releases/tag/1.4.0)

**Comment by blag** [comment]
Hey @friedcell, thanks for this PR! It made it really really easy for me to write PR #100.

**Pull request merged**